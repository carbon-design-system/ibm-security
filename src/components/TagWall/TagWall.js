/**
 * @file Tag wall.
 * @copyright IBM Security 2019 - 2021
 */

import classnames from 'classnames';
import { arrayOf, bool, func, shape, string, object } from 'prop-types';
import React from 'react';

import { carbonPrefix, getComponentNamespace } from '../../globals/namespace';

import * as defaultLabels from '../../globals/nls';

import defaultItemToString from '../__tools__/defaultItemToString';

import Button from '../Button';
import InteractiveTag from '../InteractiveTag';

const namespace = getComponentNamespace('tag-wall');
const noop = () => {};

export const type = 'UNSELECT_ITEM';

/**
 * Tag wall component.
 */
const TagWall = ({
  addButtonDisabled,
  addLabel,
  className,
  disable,
  items,
  itemToString,
  label,
  onAddButton,
  onChange,
  labels,
  ...other
}) => {
  const { TAG_WALL_ADD_BUTTON, TAG_WALL_LABEL, TAG_WALL_REMOVE_BUTTON } = {
    ...defaultLabels.labels,
    ...defaultLabels.filterFalsey({
      TAG_WALL_ADD_BUTTON: addLabel,
      TAG_WALL_LABEL: label,
    }),
    ...labels,
  };

  return (
    <section
      className={classnames(namespace, className)}
      aria-label={TAG_WALL_LABEL}
      {...other}
    >
      {TAG_WALL_LABEL && (
        <h2 className={`${namespace}__label`}>{TAG_WALL_LABEL}</h2>
      )}

      {items.map((item, index) => {
        const key = `tag__${index}`;

        return (
          <InteractiveTag
            id={item.id}
            key={key}
            isSelected={item.isSelected}
            onRemove={event => {
              event.stopPropagation();

              onChange({ item, type });
            }}
            removable={!disable}
            removeBtnLabel={TAG_WALL_REMOVE_BUTTON}
            type="gray"
            {...item.props}
          >
            <span
              className={`${carbonPrefix}--text-truncate--end`}
              title={itemToString(item)}
            >
              {itemToString(item)}
            </span>
          </InteractiveTag>
        );
      })}
      {!addButtonDisabled && (
        <Button
          disabled={disable}
          kind="ghost"
          onClick={onAddButton}
          size="small"
        >
          {TAG_WALL_ADD_BUTTON}
        </Button>
      )}
    </section>
  );
};

TagWall.propTypes = {
  /** @type {boolean} Determines whether or not the 'add' button is disabled. */
  addButtonDisabled: bool,

  /** @type {string} Label for button that adds tag. */
  addLabel: string,

  /** @type {string} Extra classes to add. */
  className: string,

  /** @type {boolean} Determines whether or not tag management is enabled. */
  disable: bool,

  /** @type {Array.<Object.<string,boolean>>} List of tags to be consumed. */
  items: arrayOf(
    shape({
      id: string.isRequired,
      isSelected: bool,
      label: string.isRequired,
      props: object,
    })
  ).isRequired,

  /** @type {function} Function that converts object to string. */
  itemToString: func,

  /** @type {string} Description label. */
  label: string,

  /** @type {Object.<string, *>} Default translation labels object. */
  labels: defaultLabels.propType,

  /** @type {function} Handles `onClick` for the 'add' button. */
  onAddButton: func,

  /** @type {function} Communicates to the consuming component any changes to the tags. */
  onChange: func,
};

TagWall.defaultProps = {
  addButtonDisabled: false,
  addLabel: '',
  className: '',
  disable: false,
  itemToString: defaultItemToString,
  label: null,
  labels: {},
  onAddButton: noop,
  onChange: noop,
};

export default TagWall;
