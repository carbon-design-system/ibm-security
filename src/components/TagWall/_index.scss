////
/// Tag wall.
/// @group tag-wall
/// @copyright IBM Security 2019
////

@import '../Component/mixins';

@import 'mixins';

@import '../Tag/index';

@include security--component($name: tag-wall) {
  @include tag-wall;
}
