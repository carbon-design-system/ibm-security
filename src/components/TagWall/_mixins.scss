////
/// Tag wall mixins.
/// @group tag-wall
/// @copyright IBM Security 2019 - 2021
////

@import '@carbon/type/scss/index';

@mixin tag-wall {
  &__label {
    $tag-wall__spacing: $carbon--spacing-03;

    @include carbon--type-style($name: caption-01);
    margin-top: $tag-wall__spacing;
    margin-bottom: $tag-wall__spacing;
  }

  > #{$tag--interactive__namespace} {
    max-width: 40ex;

    &__button {
      flex-shrink: 0;
    }
  }
}
