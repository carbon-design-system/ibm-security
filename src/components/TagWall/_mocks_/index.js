/**
 * @file Tag wall mocks
 * @copyright IBM Security 2019
 */

export default {
  items: new Array(8).fill().map((value = 'Tag', id) => ({
    id: `${id}`,
    label: `${value} ${id + 1}`,
    isSelected: id < 3,
    props: {
      [`example-attribute`]: 'example-value',
    },
  })),
  label: 'Label',
  addLabel: 'Add',
};
