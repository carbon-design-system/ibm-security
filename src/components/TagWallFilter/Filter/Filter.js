/**
 * @file Filter.
 * @copyright IBM Security 2019
 */

import Add20 from '@carbon/icons-react/lib/add/20';
import Search20 from '@carbon/icons-react/lib/search/20';

import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import Downshift from 'downshift';
import ListBox from 'carbon-components-react/lib/components/ListBox';
import { defaultFilterItems } from 'carbon-components-react/lib/components/ComboBox/tools/filter';

import theme from '../../../globals/theme';

import { defaultItemToString } from './tools/itemToString';
import { defaultSortItems, defaultCompareItems } from './tools/sorting';

import Icon from '../../Icon';

import {
  carbonPrefix,
  getComponentNamespace,
} from '../../../globals/namespace/index';

const { Field, Menu, MenuItem, Selection } = ListBox;

const namespace = getComponentNamespace('filter');

class Filter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      highlightedIndex: null,
      inputValue: '',
    };
  }

  handleOnInputKeyDown = event => {
    event.stopPropagation();
  };

  handleOnInputValueChange = (inputValue, stateAndHelpers) => {
    if (stateAndHelpers.type === '__autocomplete_mouseup__') return;
    this.setState(() => {
      if (Array.isArray(inputValue)) {
        return {
          inputValue: '',
        };
      }

      return {
        inputValue: inputValue || '',
      };
    });
  };

  clearInputValue = event => {
    event.stopPropagation();
    this.setState({ inputValue: '' });
  };

  handleOnStateChange = changes => {
    const { type } = changes;
    switch (type) {
      case Downshift.stateChangeTypes.changeInput:
        this.setState({ inputValue: changes.inputValue });
        break;
      case Downshift.stateChangeTypes.keyDownArrowDown:
      case Downshift.stateChangeTypes.keyDownArrowUp:
      case Downshift.stateChangeTypes.itemMouseEnter:
        this.setState({ highlightedIndex: changes.highlightedIndex });
        break;

      default:
    }
  };

  handleClearSelection = () => {
    this.setState({ inputValue: '' });
    this.props.onChange({ selectedItems: [], type: 'CLEAR_SELECTED_ITEMS' });
  };

  handleItemChange = selectedItem => {
    this.setState(
      { inputValue: '' },
      this.props.onChange({ item: selectedItem, type: 'SELECT_ITEM' })
    );
  };

  handleKeyDownChange = (event, item) => {
    if (event.key === 'Enter') {
      this.handleItemChange(item);
    }
  };

  buildMenuItems = getItemProps => {
    const {
      items,
      itemToString,
      filterItems,
      locale,
      selectedItems,
      sortItems,
      compareItems,
    } = this.props;

    const { inputValue } = this.state;

    const menuItems = sortItems(
      filterItems(items, { itemToString, inputValue }),
      {
        selectedItems,
        itemToString,
        compareItems,
        locale,
      }
    ).map(item => {
      const itemProps = getItemProps({ item });
      const itemText = itemToString(item);

      return (
        <MenuItem
          className={`${namespace}__list-item`}
          key={itemProps.id}
          isActive={selectedItems.indexOf(item) !== -1}
          onKeyDown={e => this.handleKeyDownChange(e, item)}
          role="button"
          tabIndex="0"
          {...itemProps}
        >
          <div
            className={`${namespace}__list-item__entry`}
            aria-labelledby={itemProps.id}
          >
            {itemText}
            <span className={`${namespace}__add`}>
              <Icon className={`${namespace}__add__icon`} renderIcon={Add20} />
            </span>
          </div>
        </MenuItem>
      );
    });

    return menuItems;
  };

  render() {
    const { inputValue, highlightedIndex } = this.state;
    const { itemToString, id, disabled, placeholder } = this.props;

    const clearButton = inputValue ? (
      <Selection
        clearSelection={this.clearInputValue}
        translateWithId={id => {
          switch (id) {
            case 'clear.all':
              return this.props.filterFieldClearAllTooltip;
            case 'clear.selection':
              return this.props.filterFieldClearSelectionTooltip;
            default:
              return '';
          }
        }}
      />
    ) : null;

    return (
      <Downshift
        highlightedIndex={highlightedIndex}
        inputValue={inputValue}
        onInputValueChange={this.handleOnInputValueChange}
        onChange={this.handleItemChange}
        itemToString={itemToString}
        onStateChange={this.handleOnStateChange}
        clearSelection={this.handleClearSelection}
        isOpen
        render={({
          getButtonProps,
          getInputProps,
          getItemProps,
          getRootProps,
        }) => (
          <ListBox
            className={namespace}
            disabled={disabled}
            {...getRootProps({ refKey: 'innerRef' })}
            id={id}
          >
            <Field {...getButtonProps({ disabled, tabIndex: -1 })} id={id}>
              <input
                className={`${namespace}__input-field ${carbonPrefix}text-input`}
                {...getInputProps({
                  disabled,
                  id,
                  placeholder,
                  onBlur: event => event.preventDefault(),
                  onKeyDown: this.handleOnInputKeyDown,
                })}
              />
              {clearButton}
              <span className={`${namespace}__search`}>
                <Icon
                  className={classnames(`${namespace}__search__icon`, {
                    [`${namespace}__input--active`]:
                      this.state.inputValue !== '',
                  })}
                  renderIcon={Search20}
                  fill={theme.icon01}
                />
              </span>
            </Field>
            <Menu className={`${namespace}__list-container`} id={id}>
              {this.buildMenuItems(getItemProps)}
            </Menu>
          </ListBox>
        )}
      />
    );
  }
}

Filter.defaultProps = {
  className: '',
  onChange: '',
  compareItems: defaultCompareItems,
  disabled: false,
  filterItems: defaultFilterItems,
  initialSelectedItems: [],
  selectedItems: [],
  itemToString: defaultItemToString,
  locale: 'en',
  sortItems: defaultSortItems,
  filterFieldClearSelectionTooltip: 'Clear selected item',
  filterFieldClearAllTooltip: 'Clear all selected items',
};

Filter.propTypes = {
  /** @type {string} Extra classes to add. */
  className: PropTypes.string,

  /** @type {arrayOf} Collection of data structure. */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
    })
  ).isRequired,

  /** @type {arrayOf} Arbitrary items from their collection */
  initialSelectedItems: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
    })
  ),
  selectedItems: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
    })
  ),

  /** @type {func} Render a given item to a string label */
  itemToString: PropTypes.func,

  /** @type {string} Used for sorting the list of items */
  locale: PropTypes.string,

  /** @type {func} Used for internal state changes that are occuring */
  onChange: PropTypes.func,

  /** @type {string} Used as the textual representation */
  placeholder: PropTypes.string.isRequired,

  /** @type {Function} Provide a compare function that is used to determine the ordering of options. */
  compareItems: PropTypes.func,

  /** @type {func} Handle the sort logic for selected versus un-selected items. */
  sortItems: PropTypes.func,

  /** @type {string} Specify a custom id. */
  id: PropTypes.string.isRequired,

  /** @type {func} Handle the filter of items. */
  filterItems: PropTypes.func,

  /** @type {bool} Handle the disabled of selected items. */
  disabled: PropTypes.bool,

  /** @type {string} tooltip label for clearing a selected item */
  filterFieldClearSelectionTooltip: PropTypes.string,

  /** @type {string} tooltip label for clearing all selected items */
  filterFieldClearAllTooltip: PropTypes.string,
};

export default Filter;
