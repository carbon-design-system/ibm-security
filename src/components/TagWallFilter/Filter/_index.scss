////
/// Filter component.
/// @group filter
/// @copyright IBM Security 2019
////

@import 'carbon-components/scss/components/multi-select/multi-select';

@import '../../Component/mixins';
@import 'mixins';

@include security--component($name: $filter__name) {
  @include filter;
}
