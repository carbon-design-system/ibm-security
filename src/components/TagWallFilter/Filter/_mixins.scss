////
/// Filter mixins.
/// @filter
/// @copyright IBM Security 2019 - 2020
////

@import '@carbon/layout/scss/spacing';
@import '@carbon/layout/scss/mini-unit';
@import '@carbon/themes/scss/tokens';

@import 'carbon-components/scss/globals/scss/helper-classes';
@import 'carbon-components/scss/globals/scss/helper-mixins';

@import '../../../globals/namespace/index';

@import '../../Tearsheet/mixins';

/// Name.
/// @type String
$filter__name: filter;

/// Namespace.
/// @type String
$filter__namespace: get-component-namespace(
  $component-name: $filter__name,
);

/// Dimensions.
/// @type Length
$filter__sizing__dimensions: carbon--mini-units(
  $count: 5,
);

@mixin filter {
  $root: &;

  .#{$prefix}--list-box {
    &__field {
      position: fixed;
      height: $filter__sizing__dimensions;
      padding-right: 0;
      padding-left: 0;
      background-color: inherit;
      transform: translate3d(0, -100%, 0);
    }

    &__menu-item__option {
      display: flex;
      align-items: center;
      justify-content: space-between;
    }
  }

  &__search {
    display: flex;
    align-items: center;
    position: absolute;
    right: $carbon--spacing-05;
    height: 100%;
    top: 0;

    &__icon {
      fill: $icon-02;
    }
  }

  &__add {
    display: flex;

    &__icon {
      fill: currentColor;
      visibility: hidden;
    }
  }

  &__list-container {
    background-color: $tearsheet__color__background;

    &:empty {
      min-height: $filter__sizing__dimensions;
      cursor: default;
    }
  }

  &__list-item {
    &:focus,
    &:hover {
      #{$root}__add__icon {
        visibility: visible;
      }
    }

    &:focus {
      @include focus-outline($type: outline);
    }

    &:hover {
      background: $ui-02;
      position: relative;

      &:after {
        content: '';
        position: absolute;
        width: 100%;
        background: $ui-02;
        height: 1px;
        bottom: 0;
        transform: translateY(1px);
      }
    }

    &:first-of-type > .#{$prefix}--list-box__menu-item__option {
      border-color: transparent;
    }
  }

  &__list-item__entry {
    align-items: center;
    width: 100%;
    justify-content: space-between;
    display: flex;
  }
}
