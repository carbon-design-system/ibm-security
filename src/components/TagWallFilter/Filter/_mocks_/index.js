/**
 * @file Filter mocks.
 * @copyright IBM Security 2018
 */

const items = [
  {
    id: 'item-1',
    label: 'New York',
  },
  {
    id: 'item-2',
    label: 'London',
  },
  {
    id: 'item-3',
    label: 'Tokyo',
  },
  {
    id: 'item-4',
    label: 'Beijing',
  },
  {
    id: 'item-5',
    label: 'New Delhi',
  },
  {
    id: 'item-6',
    label: 'Berlin',
  },
  {
    id: 'item-7',
    label: 'Montreal',
  },
  {
    id: 'item-8',
    label: 'San Fransisco',
  },
  {
    id: 'item-9',
    label: 'Rio de Janeiro',
  },
  {
    id: 'item-10',
    label: 'Johannesburg',
  },
  {
    id: 'item-11',
    label: 'Abuja',
  },
];

const itemToString = (item) => (item ? item.label : '');

export { items, itemToString };
