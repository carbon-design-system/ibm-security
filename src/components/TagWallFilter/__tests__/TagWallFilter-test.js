import React from 'react';
import { shallow, mount } from 'enzyme';
import {
  selectedItemsReducer,
  availableItemsReducer,
  withItemReducer,
  FilterTagFragment,
  noop,
  TagWallFilter,
  itemToString,
} from '../TagWallFilter';
import TagWall from '../../TagWall';
import Filter from '../Filter';

describe('TagWallFilter tests', () => {
  const tearsheetProps = {
    heading: 'TagWallFilter Heading',
    description:
      'Whatâ€™s the secret to minced and ground doughnut? Always use small garlic.',
    closeButton: {
      onClick: jest.fn(),
    },
    secondaryButton: {
      onClick: jest.fn(),
    },
    primaryButton: {
      onClick: jest.fn(),
    },
  };

  it('should do nothing on noop', () => {
    expect(noop()).toBeUndefined();
  });
  describe('selectedItemsReducer', () => {
    let initState;
    beforeEach(() => {
      initState = { items: [] };
    });
    it('should handle SELECT_ITEM', () => {
      const state1 = selectedItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'x', label: 'hallo' },
      });
      const state2 = selectedItemsReducer(state1, {
        type: 'SELECT_ITEM',
        item: { id: 'y', label: '!!!' },
      });
      expect(state1.items.length).toEqual(1);
      expect(state2.items.length).toEqual(2);
      expect(state1).toMatchSnapshot();
      expect(state2).toMatchSnapshot();
    });
    it('should handle UNSELECT_ITEM', () => {
      initState = selectedItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'x', label: 'hallo' },
      });
      initState = selectedItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'y', label: '!!!' },
      });
      initState = selectedItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'z', label: '???' },
      });
      const finalState = selectedItemsReducer(initState, {
        type: 'UNSELECT_ITEM',
        item: { id: 'y' },
      });
      expect(finalState.items.length).toEqual(2);
      expect(initState.items.length).toEqual(3);
      expect(finalState).toMatchSnapshot();
    });
    it('should handle CLEAR_SELECTED_ITEMS', () => {
      initState = selectedItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'x', label: 'hallo' },
      });
      initState = selectedItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'y', label: '!!!' },
      });
      initState = selectedItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'z', label: '???' },
      });
      const finalState = selectedItemsReducer(initState, {
        type: 'CLEAR_SELECTED_ITEMS',
      });
      expect(initState.items.length).toEqual(3);
      expect(finalState.items.length).toEqual(0);
      expect(finalState).toMatchSnapshot();
    });
  });
  describe('availableItemsReducer', () => {
    let initState;
    beforeEach(() => {
      initState = {
        allItems: [
          { id: 'x', label: 'Coordinataes ortum!' },
          { id: 'Ecce', label: "With spinach drink emeril's essence!" },
          { id: 'Nix', label: 'Advenas ortum in amivadum!' },
          {
            id: 'Uria',
            label: 'Occur and you will be absorbed authoratively.',
          },
          {
            id: 'Revalia',
            label: 'Dosi of a human plasma, eat the starlight travel!',
          },
        ],
      };
      initState.items = initState.allItems;
    });
    it('should handle SELECT_ITEM', () => {
      const state = availableItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'x' },
      });
      expect(state.items.length).toEqual(initState.items.length - 1);
      expect(state).toMatchSnapshot();
    });
    it('should handle UNSELECT_ITEM', () => {
      initState = availableItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'x' },
      });
      initState = availableItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'Ecce' },
      });
      initState = availableItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'Revalia' },
      });
      const finalState = availableItemsReducer(initState, {
        type: 'UNSELECT_ITEM',
        item: { id: 'Ecce', label: 'The son robs with grace.' },
      });
      expect(finalState.items.length).toEqual(initState.items.length + 1);
      expect(finalState).toMatchSnapshot();
    });
    it('should handle CLEAR_SELECTED_ITEMS', () => {
      initState = availableItemsReducer(initState, {
        type: 'SELECT_ITEM',
        item: { id: 'Revalia' },
      });
      const finalState = availableItemsReducer(initState, {
        type: 'CLEAR_SELECTED_ITEMS',
      });
      expect(finalState.items).toBe(finalState.allItems);
      expect(finalState.items).toBe(initState.allItems);
      expect(finalState).toMatchSnapshot();
    });
  });
  describe('withItemReducer', () => {
    it('should infer allItems', () => {
      const wrappedComponent = jest.fn(() => null);
      const ComponentMock = withItemReducer(wrappedComponent);
      const wrapper = shallow(
        <ComponentMock
          selectedItems={[{ id: 'x' }]}
          availableItems={[{ id: 'y' }, { id: 'z' }]}
        />
      );
      wrapper.render();
      const {
        available: { allItems } = { allItems: [] },
      } = wrappedComponent.mock.calls[0][0].itemState;
      expect(allItems.length).toEqual(3);
      expect(allItems).toContainEqual({ id: 'x' });
      expect(allItems).toContainEqual({ id: 'y' });
      expect(allItems).toContainEqual({ id: 'z' });
    });
  });
  describe('FilterTagFragment', () => {
    it('should render the fragment', () => {
      const onChangeMock = jest.fn();
      const wrapper = mount(
        <FilterTagFragment
          id="test-id"
          onChange={onChangeMock}
          selectedItems={[{ id: 'y', label: 'Y' }]}
          availableItems={[{ id: 'x', label: 'X' }]}
          placeholder="placeholder"
          itemToString={itemToString}
          tearsheetProps={tearsheetProps}
        />
      );

      const tagWallWrapper = wrapper.find(TagWall);

      expect(wrapper).toMatchSnapshot();
      expect(tagWallWrapper).toMatchSnapshot();
    });
    it('should call the on change callback with the latest item state on change', () => {
      const onChangeMock = jest.fn();
      const wrapper = shallow(
        <FilterTagFragment
          id="test-id"
          onChange={onChangeMock}
          selectedItems={[{ id: 'y', label: 'Y' }]}
          availableItems={[{ id: 'x', label: 'X' }]}
          placeholder="placeholder"
          itemToString={itemToString}
          tearsheetProps={tearsheetProps}
        />
      );

      const filterWrapper = wrapper
        .first()
        .shallow()
        .first()
        .shallow()
        .find(Filter)
        .first();
      filterWrapper.prop('onChange')({
        type: 'SELECT_ITEM',
        item: { id: 'x' },
      });
      expect(wrapper.props()).toMatchSnapshot();
      expect(onChangeMock).toHaveBeenCalled();
      expect(onChangeMock).toMatchSnapshot();
      onChangeMock.mockClear();
      filterWrapper.prop('onChange')({
        type: 'UNSELECT_ITEM',
        item: { id: 'x' },
      });

      expect(wrapper.props()).toMatchSnapshot();
      expect(onChangeMock).toHaveBeenCalled();
      expect(onChangeMock).toMatchSnapshot();
    });
  });
  describe('TagWallFilter', () => {
    it('should mount the TagWallFilter component', () => {
      const wrapper = mount(
        <TagWallFilter
          id="test-id"
          selectedItems={[{ id: 'x', label: 'X' }]}
          availableItems={[{ id: 'y', label: 'Y' }]}
          {...tearsheetProps}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
