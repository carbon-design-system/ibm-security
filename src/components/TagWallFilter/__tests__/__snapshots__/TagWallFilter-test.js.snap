// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagWallFilter calls \`onChange\` when items are selected and unselected 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "available": Object {
          "allItems": undefined,
          "items": Array [
            Object {
              "id": "0",
              "label": "TagWallFilter 0",
            },
          ],
        },
        "selected": Object {
          "items": Array [
            Object {
              "id": "1",
              "label": "TagWallFilter 1",
            },
          ],
        },
      },
    ],
    Array [
      Object {
        "available": Object {
          "allItems": Array [
            Object {
              "id": "0",
              "label": "TagWallFilter 0",
            },
            Object {
              "id": "1",
              "label": "TagWallFilter 1",
            },
          ],
          "items": Array [],
        },
        "selected": Object {
          "items": Array [
            Object {
              "id": "0",
              "isSelected": true,
              "label": "TagWallFilter 0",
            },
            Object {
              "id": "1",
              "label": "TagWallFilter 1",
            },
          ],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`TagWallFilter calls \`onChange\` when items are selected and unselected 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "available": Object {
          "allItems": Array [
            Object {
              "id": "0",
              "label": "TagWallFilter 0",
            },
            Object {
              "id": "1",
              "label": "TagWallFilter 1",
            },
          ],
          "items": Array [
            Object {
              "id": "1",
              "label": "TagWallFilter 1",
            },
          ],
        },
        "selected": Object {
          "items": Array [
            Object {
              "id": "0",
              "isSelected": true,
              "label": "TagWallFilter 0",
            },
          ],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`TagWallFilter renders 1`] = `
<body
  class="security--portal__container"
>
  <div>
    <span
      class="security--transition"
    />
  </div>
  <section
    aria-hidden="false"
    class="security--tearsheet--small security--tag-wall-filter"
  >
    <header
      class="security--tearsheet--small__header"
    >
      <h1
        class="security--tearsheet--small__title"
      >
        TagWallFilter
      </h1>
      <div
        class="security--tearsheet--small__description"
      >
        <div
          class="security--tag-wall-filter__description"
        >
          
          <section
            aria-label=""
            class="security--tag-wall security--tag-wall-filter__tag-wall"
          >
            
            <div
              aria-label=" TagWallFilter 1"
              class="bx--tag security--tag--interactive security--tag--interactive--default bx--tag--filter bx--tag--gray"
              id="1"
            >
              <span
                class="bx--tag__label"
                title="TagWallFilter 1"
              >
                TagWallFilter 1
              </span>
              <button
                aria-labelledby="1"
                class="bx--tag__close-icon"
                title=""
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 32 32"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
                  />
                </svg>
              </button>
            </div>
          </section>
        </div>
      </div>
    </header>
    <section
      class="security--tearsheet--small__body"
    >
      <div
        class="security--scroll-gradient security--scroll-gradient--none security--scroll-gradient--y security--tearsheet--small__content"
        role="presentation"
      >
        <div
          aria-hidden="true"
          class="security--scroll-gradient__before"
          role="presentation"
        />
        <div
          class="security--scroll-gradient__content"
        >
          <div
            class=""
          >
            <div
              aria-expanded="true"
              aria-haspopup="listbox"
              aria-labelledby="downshift-0-label"
              aria-owns="downshift-0-menu"
              class="security--filter bx--list-box"
              id="security--tag-wall-filter"
              role="combobox"
            >
              <div
                aria-haspopup="true"
                aria-label="close menu"
                class="bx--list-box__field"
                data-toggle="true"
                id="security--tag-wall-filter"
                role="button"
                tabindex="-1"
                type="button"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="downshift-0-menu"
                  aria-labelledby="downshift-0-label"
                  autocomplete="off"
                  class="security--filter__input-field bx--text-input"
                  id="security--tag-wall-filter"
                  placeholder=""
                  value=""
                />
                <span
                  class="security--filter__search"
                >
                  <svg
                    aria-hidden="true"
                    class="security--icon security--filter__search__icon"
                    fill="var(--cds-icon-01, #f4f4f4)"
                    focusable="false"
                    height="20"
                    preserveAspectRatio="xMidYMid meet"
                    style="will-change: transform;"
                    viewBox="0 0 32 32"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M29,27.5859l-7.5521-7.5521a11.0177,11.0177,0,1,0-1.4141,1.4141L27.5859,29ZM4,13a9,9,0,1,1,9,9A9.01,9.01,0,0,1,4,13Z"
                    />
                  </svg>
                </span>
              </div>
              <div
                class="security--filter__list-container"
                id="security--tag-wall-filter"
                role="listbox"
              >
                <div
                  aria-selected="false"
                  class="bx--list-box__menu-item"
                  id="downshift-0-item-0"
                  role="option"
                  tabindex="0"
                >
                  <div
                    class="bx--list-box__menu-item__option"
                  >
                    <div
                      aria-labelledby="downshift-0-item-0"
                      class="security--filter__list-item__entry"
                    >
                      <span
                        class="bx--text-truncate--end"
                        title="TagWallFilter 0"
                      >
                        TagWallFilter 0
                      </span>
                      <span
                        class="security--filter__add"
                      >
                        <svg
                          aria-hidden="true"
                          class="security--icon security--filter__add__icon"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          style="will-change: transform;"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
                          />
                        </svg>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-hidden="true"
          class="security--scroll-gradient__after"
          role="presentation"
        />
      </div>
    </section>
    <footer
      class="security--tearsheet--small__footer"
    >
      <button
        class="security--tearsheet--small__footer__button security--tearsheet--small__footer__button--secondary bx--btn bx--btn--xl bx--btn--secondary"
        tabindex="0"
        type="button"
      >
        Cancel
      </button>
      <button
        class="security--tearsheet--small__footer__button bx--btn bx--btn--xl bx--btn--primary"
        tabindex="0"
        type="button"
      >
        Save
      </button>
    </footer>
    <button
      aria-label="Close"
      class="security--button--icon security--tearsheet--small__button--close security--button--icon--lg"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="security--icon"
        fill="currentColor"
        focusable="false"
        height="20"
        preserveAspectRatio="xMidYMid meet"
        style="will-change: transform;"
        viewBox="0 0 32 32"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
        />
      </svg>
    </button>
  </section>
  <div
    class="security--portal__overlay"
    tabindex="-1"
  />
</body>
`;

exports[`TagWallFilter withItemReducer CLEAR_SELECTED_ITEMS clears selected items 1`] = `
Object {
  "available": Object {
    "allItems": Array [],
    "items": Array [],
  },
  "selected": Object {
    "items": Array [],
  },
}
`;

exports[`TagWallFilter withItemReducer SELECT_ITEM selects an item 1`] = `
Object {
  "available": Object {
    "allItems": Array [],
    "items": Array [],
  },
  "selected": Object {
    "items": Array [
      Object {
        "id": "0",
        "isSelected": true,
        "label": "TagWallFilter 0",
      },
    ],
  },
}
`;

exports[`TagWallFilter withItemReducer UNSELECT_ITEM unselects an item 1`] = `
Object {
  "available": Object {
    "allItems": Array [],
    "items": Array [
      Object {
        "id": "0",
        "label": "TagWallFilter 0",
      },
    ],
  },
  "selected": Object {
    "items": Array [],
  },
}
`;
